gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js

```diff
@@ -35,45 +43,197 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
     }
     var url = fluid.module.resolvePath(that.options.path);
     if (!fs.existsSync(url)) {
         fluid.fail("The path provided to the solutionsRegistry datasource (", url, ") has not been found on the file system");
     }
-
     that.fullSolutionsRegistry = fluid.freezeRecursive(require(url));
 };
 
+// Solutions registry data source used by the Cloud Based Flow Manager
+fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource.cloud", {
+    gradeNames: ["gpii.flowManager.solutionsRegistry.dataSource"],
+    readOnlyGrade: "gpii.flowManager.solutionsRegistry.dataSource.cloud",
+    invokers: {
+        getImpl: {
+            funcName: "gpii.flowManager.solutionsRegistry.dataSource.cloud.handle",
+            args: ["{that}", "{arguments}.1", "{arguments}.2"]
+                             // options, directModel
+        }
+    },
+    listeners: {
+        "loadSolutions.loadFromLocalDisk": {
+            listener: "{that}.loadFromLocalDisk",
+            priority: "first"
+        },
+        "loadSolutions.solutionsLoaded": {
+            listener: "{that}.events.solutionsRegistryReady",
+            priority: "after:loadFromLocalDisk"
+        }
+    }
+});
+
 /**
- * Handler for get requests of solutions registry. It will return either a full solution registry,
- * or if an 'os' is provided in the requestOptions, only the entries for that os will be returned
+ * Handler for get requests to the solutions registry used by the Cloud Based
+ * Flow Manager. It will return either an object containing solutions registries
+ * for the given platform.
  *
- * @param {Object} that - The gpii.flowManager.solutionsRegistry.dataSource.
- * @param {Object} requestOptions - Currently the only request option supported is "os". If provided,
- *           the returned solutions registry will be filtered by OS version.
- * @return {Promise} A promise that will be resolved with results (see above) or rejected on error.
+ * @param {Object} that - The gpii.flowManager.solutionsRegistry.dataSource.cloud
+ * @param {Object} requestOptions - Currently the only request option supported
+ *          is "os". The returned solutions registry will be the solutions for
+ *          the OS version provided.
+ * @return {Promise} A promise that will be resolved with results (see above) or
+ *          rejected on error.
+ */
+gpii.flowManager.solutionsRegistry.dataSource.cloud.handle = function (that, requestOptions) {
+    var promise = fluid.promise();
+    if (requestOptions.os && requestOptions.os in that.fullSolutionsRegistry) {
+        promise.resolve(that.fullSolutionsRegistry[requestOptions.os]);
+    } else {
+        promise.reject({
+            isError: true,
+            message: "The requested OS (" + requestOptions.os + ") was not present in the solutions registry",
+            statusCode: 404
+        });
+    }
+    return promise;
+};
+
+// Solutions registry datasource used by the Local Flow Manager.  It loads both
```

Comment:
It would be nice if you can expand on the second sentence with the workflow and fallback. Thanks.

Reply:
Done.